src/stages/sender_recovery.rs

```diff
@@ -0,0 +1,300 @@
+use crate::{
+    accessors::chain,
+    stagedsync::stage::{ExecOutput, Stage, StageInput, SyncError, UnwindInput},
+    MutableTransaction, StageId,
+};
+use async_trait::async_trait;
+use ethereum::{Transaction, TransactionMessage};
+use ethereum_types::Address;
+use secp256k1::{
+    recovery::{RecoverableSignature, RecoveryId},
+    Message, SECP256K1,
+};
+use sha3::{Digest, Keccak256};
+use std::cmp;
+use tracing::error;
+
+const BUFFER_SIZE: u64 = 5000;
+
+fn recover_sender(tx: &Transaction) -> anyhow::Result<Address> {
+    let mut sig = [0u8; 64];
+    sig[..32].copy_from_slice(tx.signature.r().as_bytes());
+    sig[32..].copy_from_slice(tx.signature.s().as_bytes());
+
+    let rec = RecoveryId::from_i32(tx.signature.standard_v() as i32)?;
+
+    let public = &SECP256K1.recover(
+        &Message::from_slice(TransactionMessage::from(tx.clone()).hash().as_bytes())?,
+        &RecoverableSignature::from_compact(&sig, rec)?,
+    )?;
+
+    let address_slice = &Keccak256::digest(&public.serialize_uncompressed()[1..])[12..];
+    Ok(Address::from_slice(address_slice))
+}
+
+async fn process_block<'tx, RwTx>(tx: &mut RwTx, height: u64) -> anyhow::Result<()>
+where
+    RwTx: MutableTransaction<'tx>,
+{
+    let hash = chain::canonical_hash::read(tx, height)
+        .await?
+        .ok_or(SyncError::HashNotFound(height))?;
+    let body = chain::storage_body::read(tx, hash, height)
+        .await?
+        .ok_or(SyncError::BlockBodyNotFound(height))?;
+    let txs = chain::tx::read(tx, body.base_tx_id, body.tx_amount).await?;
+
+    let mut senders = vec![];
+    for tx in &txs {
+        senders.push(recover_sender(tx)?);
+    }
+
+    chain::tx_sender::write(tx, body.base_tx_id, &senders).await
+}
+
+#[derive(Debug)]
+pub struct SenderRecovery;
+
+#[async_trait]
+impl<'db, RwTx> Stage<'db, RwTx> for SenderRecovery
+where
+    RwTx: MutableTransaction<'db>,
+{
+    fn id(&self) -> StageId {
+        StageId("SenderRecovery")
+    }
+
+    fn description(&self) -> &'static str {
+        "Recovering senders of transactions from signatures"
+    }
+
+    async fn execute<'tx>(&self, tx: &'tx mut RwTx, input: StageInput) -> anyhow::Result<ExecOutput>
+    where
+        'db: 'tx,
+    {
+        let from_height = input.stage_progress.unwrap_or(0);
+        let max_height = if let Some((_, height)) = input.previous_stage {
+            height
+        } else {
+            0
+        };
+        let to_height = cmp::min(max_height, from_height + BUFFER_SIZE);
+
+        let mut height = from_height;
+        while height < to_height {
+            if let Err(e) = process_block(tx, height + 1).await {
+                error!(
+                    "Failed to recover senders for block {}: {}",
+                    height + 1,
+                    e.to_string(),
+                );
+                break;
+            }
```

Comment:
```suggestion
            process_block(tx, height + 1).await.with_context(|| {
                format!("Failed to recover senders for block {}"
            }))?;
```

Stages must not handle errors, let alone suppress them.